### PR TITLE
[MRG] ENH: plot_input() accepts flexible spike_types argument

### LINKED ISSUE #135: plot_input() should plot all inputs
The `Network.plot_input()` method currently only plots evoked inputs. It should plot all input types (e.g., rhythmic, Poisson, etc.). See https://github.com/jonescompneurolab/hnn/issues/178#issuecomment-671624649.

### PR DESCRIPTION
**Pull request by ntolley**
closes #135. Moves `network.plot_input()` to `network.spikes.plot_input()` (tests and examples updated to reflect this). 

Adds new argument `spike_types` to function call. Examples calls included below:

`spikes.plot_input(spike_types=['evprox1', 'evdist1','evprox2']`
image

`spikes.plot_input(spike_types={'Early': ['evprox1', 'evdist1'], 'Late': ['evprox2']})`
image

`spikes.plot_input(spike_types={'Evoked': ['ev']})`
image

### CODE REVIEW

File: hnn_core/network.py [hunk 1]
```diff
@@ -484,6 +466,36 @@ def plot(self, ax=None, show=True):
...
+            String input of a valid spike type is plotted individually.
+                Ex: 'common', 'evdist', 'evprox', 'extgauss', 'extpois'
+            List of valid string inputs will plot each spike type individually.
+                Ex: ['common', 'evdist']
+            Dictionary of valid lists will plot list elements as a group.
+                Ex: {'Evoked': ['evdist', 'evprox'], 'External': ['extpois']}
+            If None, all default spike types are plotted individually.
+                Default: ['common', 'evdist', 'evprox', 'extgauss', 'extpois']
+            Valid strings also include leading characters of spike types
+                    Example: 'ext' is equivalent to ['extgauss', 'extpois']
```
Comment (jasmainak) [inline hnn_core/network.py:487]: indentation issue

File: hnn_core/tests/test_network.py [hunk 2]
```diff
@@ -63,6 +63,7 @@ def test_spikes():
     spikegids = [[1, 3], [5, 7]]
     spiketypes = [['L2_pyramidal', 'L2_basket'], ['L5_pyramidal', 'L5_basket']]
     spikes = Spikes(times=spiketimes, gids=spikegids, types=spiketypes)
+    spikes.plot_input(show=False)
```
Comment (jasmainak) [inline hnn_core/tests/test_network.py:66]: tests could be more extensive. We could check for e.g., that errors are raised if `spike_types` are not of the right type

File: hnn_core/viz.py [hunk 3]
```diff
@@ -64,20 +77,37 @@ def plot_hist_input(net, ax=None, show=True):
...
-    bins = np.linspace(0, net.params['tstop'], 50)
+    spike_times = np.array(sum(spikes._times, []))
+    spike_types_data = np.array(sum(spikes._types, []))
+    default_types = ['evprox', 'evdist', 'common', 'extgauss', 'extpois']
+    cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
+    input_types = np.setdiff1d(np.unique(spike_types_data), cell_types)
+    spike_types_mask = {s_type: np.in1d(spike_types_data, s_type)
+                        for s_type in input_types}
+
+    if type(spike_types) is str:
```
Comment (jasmainak) [inline hnn_core/viz.py:88]: use `isinstance`

File: hnn_core/viz.py [hunk 4]
```diff
@@ -64,20 +77,37 @@ def plot_hist_input(net, ax=None, show=True):
...
+        spike_types = {spike_types: [spike_types]}
+    elif type(spike_types) is list:
+        spike_types = {s_type: [s_type] for s_type in spike_types}
+    elif spike_types is None:
+        spike_types = {s_type: [s_type] for s_type in default_types}
```
Comment (jasmainak) [inline hnn_core/viz.py:93]: raise error if none of the above

File: hnn_core/viz.py [hunk 5]
```diff
@@ -64,20 +77,37 @@ def plot_hist_input(net, ax=None, show=True):
...
+
+    spike_mask = {s_label: np.logical_or.reduce(
+        np.logical_or.reduce([
+            [s_mask for (s_key, s_mask) in spike_types_mask.items()
+                if s_key.startswith(s_type)]
+            for s_type in s_list]))
+        for (s_label, s_list) in spike_types.items()}
```
Comment (jasmainak) [inline hnn_core/viz.py:100]: oh boy ...

I was going to comment that I was impressed by your clean code

can you simplify this double list comprehension into a simple nested for loop?

Comment (ntolley) [reply hnn_core/viz.py]: Triple comprehension is not a proud moment for me...

I've unpacked this into [for loops,](https://github.com/jonescompneurolab/hnn-core/blob/b802421711cf4e8e6af3e2ad37a814575a221f65/hnn_core/viz.py#L98) but it is somewhat unweildy. I'm not totally satisfied with the variable names but hopefully it is more legible.

File: hnn_core/tests/test_network.py (hunk 2 above)
Comment (ntolley) [reply hnn_core/tests/test_network.py:66]: I've added a [`TypeError`](https://github.com/jonescompneurolab/hnn-core/blob/b802421711cf4e8e6af3e2ad37a814575a221f65/hnn_core/viz.py#L96) check along with a matching [test](https://github.com/jonescompneurolab/hnn-core/blob/b802421711cf4e8e6af3e2ad37a814575a221f65/hnn_core/tests/test_network.py#L82). 

Do you think setting up a `ValueError` is necessary? Currently if none of the `spike_types` match the function will return a blank plot. I think this behavior makes since considering a `Spikes` object could be passed where there are no input spikes.

File: hnn_core/network.py [hunk 6]
```diff
@@ -484,6 +466,36 @@ def plot(self, ax=None, show=True):
         """
         return plot_spikes_raster(spikes=self, ax=ax, show=show)
 
+    def plot_input(self, ax=None, spike_types=None, show=True):
```
Comment (jasmainak) [inline hnn_core/network.py:469]: Do you think it can be made to work for any cell type? Or is there some fundamental issue to make it work?

File: hnn_core/viz.py [hunk 7]
```diff
@@ -64,20 +78,47 @@ def plot_hist_input(net, ax=None, show=True):
...
+    for (s_label, s_list) in spike_types.items():
+        s_label_mask = []
+        for s_type in s_list:
+            s_mask_list = []
+            for (s_key, s_mask) in spike_types_mask.items():
+                if s_key.startswith(s_type):
+                    s_mask_list.append(s_mask)
+            s_mask_list = np.logical_or.reduce(s_mask_list)
+            s_label_mask.append(s_mask_list)
+        spike_mask[s_label] = np.logical_or.reduce(s_label_mask)
```
Comment (jasmainak) [inline hnn_core/viz.py:108]: let me do a pass over this. Don't touch your code locally, otherwise you'll get merge conflicts. Meanwhile you can mull over the question above about expanding the function to be more general

File: hnn_core/network.py (hunk 6 above)
Comment (ntolley) [reply hnn_core/network.py:469]: It should work fine for any cell type! In fact it is hard-coded to ignore spikes from pyr and basket cells [here](https://github.com/jonescompneurolab/hnn-core/blob/b802421711cf4e8e6af3e2ad37a814575a221f65/hnn_core/viz.py#L85). If you remove the lines filtering out non-inputs then `spike_types_mask` will contain an element of every unique gid in `spikes._type`.

**Comment by jasmainak** [comment]
@ntolley I pushed a commit. Not sure if it's much better but I tried to do the following:

- use fewer (?) variables and spell them out fully when possible
- don't mess with np.logical_or because it's prone to errors/bugs
- raise errors for corner cases

what remains to be done is:

- [ ] make this work for any cell type. Rename the function accordingly. `plot_hist`?
- [ ] add tests to check errors are raised for the different corner cases

you can pull what I have pushed and continue from there!

File: examples/plot_simulate_somato.py [hunk 8]
```diff
@@ -82,5 +82,5 @@
 import matplotlib.pyplot as plt
 fig, axes = plt.subplots(2, 1, sharex=True, figsize=(6, 6))
 dpl[0].plot(ax=axes[0], show=False)
-net.plot_input(ax=axes[1])
+net.spikes.plot_input(ax=axes[1])
```
Comment (jasmainak) [inline examples/plot_simulate_somato.py:85]: I think we should just create `spikes` object from the constructor `Spikes` with mock data and test it. That way we can test all the different corner cases fast.

File: hnn_core/network.py (hunk 6 above)
Comment (jasmainak) [reply hnn_core/network.py:469]: okay I let you do this :)

File: hnn_core/tests/test_network.py (hunk 2 above)
Comment (jasmainak) [reply hnn_core/tests/test_network.py:66]: see my changes. The only issue there is that spikes are not grouped nicely when `spike_types=None`, everything maps to a different color. Maybe there is a way to do it ...

File: hnn_core/viz.py [hunk 9]
```diff
@@ -4,6 +4,7 @@
 #          Sam Neymotin <samnemo@gmail.com>
 
 import numpy as np
+from itertools import cycle
```
Comment (jasmainak) [inline hnn_core/viz.py:7]: I wasn't quite sure if `cycler` is installed with `maplotlib`? Went with itertools as a safer option because it's a Python builtin.

File: hnn_core/network.py (hunk 6 above)
Comment (ntolley) [reply hnn_core/network.py:469]: My current plan is to rename the function in `Spikes` and `hnn_core/viz` to `plot_hist`. However, the default behavior with `spike_types=None` will only plot the inputs (since there is already a raster function?). Let me know if this makes sense or if you have something else in mind. 

It's simple to have everything plotted be default, but the figure gets cluttered really fast.

Comment (jasmainak) [reply hnn_core/network.py:469]: yeah that sounds reasonable to me

File: examples/plot_simulate_somato.py (hunk 8 above)
Comment (ntolley) [reply examples/plot_simulate_somato.py:85]: Not sure I understand this comment. That makes sense for the tests, but for the above example wouldn't you want to use the data that was just simulated?

File: hnn_core/tests/test_network.py (hunk 2 above)
Comment (ntolley) [reply hnn_core/tests/test_network.py:66]: I'll leave it for now. Since we're moving towards giving users more control with the API it's probably good to be transparent about all `spike_types` present. That way they can easily define their own dictionary/modifications by just looking at the default plot.

File: hnn_core/network.py [hunk 10]
```diff
@@ -484,6 +466,35 @@ def plot(self, ax=None, show=True):
...
+                Example: 'ext' is equivalent to ['extgauss', 'extpois']
+        show : bool
+            If True, show the figure.
+
+        Returns
+        -------
+        fig : instance of matplotlib Figure
+            The matplotlib figure handle.
+        """
+        return plot_hist(
```
Comment (jasmainak) [inline hnn_core/network.py:495]: I would use a more descriptive name in viz.py. Something like plot_spikes_hist ?

so it's a bit clear that it plots histograms only for spikes objects

File: examples/plot_simulate_somato.py (hunk 8 above)
Comment (jasmainak) [reply examples/plot_simulate_somato.py]: well I was thinking that to test the various errors, you need to create a spikes object by doing `Spikes(...)`. But I realized now that we don't really test anything about what the spikes object contains. Your tests look fine to me.

File: hnn_core/viz.py [hunk 11]
```diff
@@ -64,21 +77,69 @@ def plot_hist_input(net, ax=None, show=True):
...
-    mask_evdist = np.in1d(gids, valid_gids)
-    bins = np.linspace(0, net.params['tstop'], 50)
+    spike_times = np.array(sum(spikes._times, []))
+    spike_types_data = np.array(sum(spikes._types, []))
+
+    unique_types = np.unique(spike_types_data)
+    spike_types_mask = {s_type: np.in1d(spike_types_data, s_type)
+                        for s_type in unique_types}
+    cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
+    input_types = np.setdiff1d(unique_types, cell_types)
```
Comment (jasmainak) [inline hnn_core/viz.py:87]: didn't we agree that we'll make this more general for any spike type. Did you check that it works for e.g., if you provide `'L5_pyramidal'`?

**Comment by jasmainak** [comment]
@ntolley you should always rebase, not merge -- against master. Because merge creates a non-linear commit history. And we prefer to have the history ordered by pull requests rather than by time.

Can you try rebasing? It should get rid of the merge commit. If you don't manage, I can help you or we can just squash all your commits when merging (but I guess you want the effort put in via 23 commits to be reflected in the contribution map? :)

File: hnn_core/tests/test_network.py [hunk 12]
```diff
@@ -79,6 +80,22 @@ def test_spikes():
...
+    with pytest.raises(TypeError, match="spike_types should be str, "
+                                        "list, dict, or None"):
+        spikes.plot_hist(spike_types=1, show=False)
+
+    with pytest.raises(TypeError, match=r"spike_types\[ev\] must be a list\. "
+                                        r"Got int\."):
+        spikes.plot_hist(spike_types={'ev': 1}, show=False)
+
+    with pytest.raises(ValueError, match=r"Elements of spike_types must map to"
+                       r" mutually exclusive input types\. L2_basket is found"
```
Comment (jasmainak) [inline hnn_core/tests/test_network.py:92]: do you actually need to escape the stop?

Comment (ntolley) [reply hnn_core/tests/test_network.py:92]: Unfortunately yes... since the error message has conditional formatting there are some rather cryptic regex matching errors. Easiest way to handle them in general is to convert to raw strings, hence the escapes.

Comment (jasmainak) [reply hnn_core/tests/test_network.py:92]: I see, I typically just put in the first few words to avoid such hassles. You pretty much guarantee a unique match given that we have only a few error messages. But this is nice too as it's more explicit

File: hnn_core/viz.py (hunk 11 above)
Comment (ntolley) [reply hnn_core/viz.py:87]: Might be a miscommunication, this was based on our discussion [here](https://github.com/jonescompneurolab/hnn-core/pull/157#discussion_r479691175). The None condition checks if there are any input spikes and plots them by default. Otherwise it plots spikes from all cells. Defining the type works the same for everything though:
image

Were you thinking of some other behavior?

File: hnn_core/tests/test_network.py (hunk 12 above)
Comment (ntolley) [reply hnn_core/tests/test_network.py:92]: Didn't realize it didn't have to be an exact match! Good to know for the future.

File: hnn_core/viz.py [hunk 13]
```diff
@@ -64,21 +77,69 @@ def plot_hist_input(net, ax=None, show=True):
...
+        else:
+            spike_types = unique_types.tolist()
+    if isinstance(spike_types, list):
+        spike_types = {s_type: [s_type] for s_type in spike_types}
+    if isinstance(spike_types, dict):
+        for spike_label in spike_types:
+            if not isinstance(spike_types[spike_label], list):
+                raise TypeError(f'spike_types[{spike_label}] must be a list. '
+                                f'Got '
+                                f'{type(spike_types[spike_label]).__name__}.')
```
Comment (jasmainak) [inline hnn_core/viz.py:104]:
```suggestion
                                f'{type(spike_types[spike_label])')
```

do you need the "name"?

File: doc/whats_new.rst [hunk 14]
```diff
@@ -27,6 +27,8 @@ Changelog
 
 - Modify plot_dipole() to accept both lists and individual instances of Dipole object, by `Nick Tolley`_ in `#145 <https://github.com/jonescompneurolab/hnn-core/pull/145>`_
 
+- Update plot_hist_input() to plot_hist() with a flexibly defined spike_types argument, by `Nick Tolley`_ in `#157 <https://github.com/jonescompneurolab/hnn-core/pull/157>`_
```
Comment (jasmainak) [inline doc/whats_new.rst:30]: I think this should go into the API section below

Comment (jasmainak) [reply doc/whats_new.rst:30]: actually never mind, it's fine here. You also added a new feature. Perhaps modify this to reflect that:

```suggestion
- Update plot_hist_input() to plot_spikes_hist() which can plot histogram of spikes for any cell type, by `Nick Tolley`_ in `#157 <https://github.com/jonescompneurolab/hnn-core/pull/157>`_
```

File: hnn_core/viz.py (hunk 11 above)
Comment (jasmainak) [reply hnn_core/viz.py:87]: This looks nice to me. Is the part about "Otherwise it plots spikes from all cells." documented in the docstring?

Also maybe we can update the example with this plot? to show that we can do it for any cell type

and also mention that this histogram is accumulated across trials, right?

File: hnn_core/viz.py (hunk 13 above)
Comment (ntolley) [reply hnn_core/viz.py:104]: Yeah while `print(type(int))` seems to convert type objects to strings just fine, it seems to be omitted when included in an error message. (at least on my computer).

**Comment by jasmainak** [comment]
@ntolley let me know when you've addressed the last couple of comments. I think this PR is ready for merge, right?

also did you manage to get rid of the merge commit?

File: hnn_core/viz.py [hunk 15]
```diff
@@ -47,14 +48,26 @@ def plot_dipole(dpl, ax=None, layer='agg', show=True):
     return ax.get_figure()
 
 
-def plot_hist_input(net, ax=None, show=True):
-    """Plot the histogram of input.
+def plot_hist(spikes, ax=None, spike_types=None, show=True):
```
Comment (jasmainak) [inline hnn_core/viz.py:51]:
```suggestion
def plot_spikes_hist(spikes, ax=None, spike_types=None, show=True):
```

Comment (jasmainak) [reply hnn_core/viz.py:51]: oh you also need to update `api.rst`. I think we forgot to do that

File: hnn_core/viz.py [hunk 16]
```diff
@@ -64,21 +77,69 @@ def plot_hist_input(net, ax=None, show=True):
...
+    bins = np.linspace(0, spike_times[-1], 50)
+    spike_color = dict()
+    for input_type, spike_label in spike_labels.items():
+        label = "_nolegend_"
+        if spike_label not in spike_color:
+            spike_color[spike_label] = next(color_cycle)
+            label = spike_label
+
+        color = spike_color[spike_label]
+        ax.hist(spike_times[spike_types_mask[input_type]], bins,
```
Comment (jasmainak) [inline hnn_core/viz.py:139]:
```suggestion
        ax.hist(spike_times[spike_types_mask[spike_type]], bins,
```

File: hnn_core/viz.py [hunk 17]
```diff
@@ -64,21 +77,69 @@ def plot_hist_input(net, ax=None, show=True):
...
     if ax is None:
         fig, ax = plt.subplots(1, 1)
-    ax.hist(spikes[mask_evprox], bins, color='r', label='Proximal')
-    ax.hist(spikes[mask_evdist], bins, color='g', label='Distal')
+
+    color_cycle = cycle(['r', 'g', 'b', 'y', 'm', 'c'])
+
+    bins = np.linspace(0, spike_times[-1], 50)
+    spike_color = dict()
+    for input_type, spike_label in spike_labels.items():
```
Comment (jasmainak) [inline hnn_core/viz.py:132]:
```suggestion
    for spike_type, spike_label in spike_labels.items():
```

**Comment by ntolley** [comment]
Just pushed the rebase, is that what you were looking for? Definitely makes sense now in terms of a clean PR history.  

And yes I think it's ready to merge once the last few comments are sorted (will push later today). I'm pretty satisfied with the plotting behavior. Plotting all inputs by default makes sense for now but we'll see after getting feedback from users.

**Comment by jasmainak** [comment]
yes history looks clean now! Let me know once you've addressed the comments

**Comment by ntolley** [comment]
I think that's everything! Let me know if there's any remaining concerns on your end, otherwise feel free to merge the PR. I'll update the example plots in the mean time.

**Comment by jasmainak** [comment]
@ntolley all is good for merge. But you have a merge commit again. You should avoid merging or pulling to avoid this problem.

**Comment by ntolley** [comment]
Will do! Didn't realize that was the convention for merging from any branch. Either way I've pushed the rebase so we're good to go.

**Comment by jasmainak** [comment]
And merged. Congratulations @ntolley ! 🎉 🍻 

looking forward to the next merge ;-)
